docs/contributing_qmk.md

```diff
@@ -0,0 +1,59 @@
+# QMK Git Conventions, or alternatively, "Git, QMK, and You: Avoiding Common Pitfalls with Contributions"
+
+This document walks through the process of contributing to QMK, detailing some ways to make this task easier.
+
+This document assumes a few things:
+
+1. You have a GitHub account, and have [forked the qmk_firmware repository](getting_started_github.md) to your account.
+2. You've [set up your build environment](newbs_getting_started.md?id=environment-setup).
+
+## Your fork's master: Update Often, Commit Never
+
+It is highly recommended for QMK development, regardless of what is being done or where, to keep your `master` branch updated, but ***never*** commit to it. Instead, do all your changes in a development branch and issue pull requests from your branches when you're developing.
+
+To reduce the chances of merge conflicts &mdash; instances where two or more users have edited the same part of a file concurrently &mdash; keep your `master` branch relatively up-to-date, and start any new developments by creating a new branch.
+
+### Updating your master branch
+
+To keep your `master` branch updated, it is recommended to add the QMK Firmware repository as a remote repository in git. To do this, enter `git remote add upstream https://github.com/qmk/qmk_firmware.git` in your Git command line interface. To verify that the repository has been added, run `git remote -v`, which should return the following:
+
+```
+$ git remote -v
+origin  https://github.com/<your_username>/qmk_firmware.git (fetch)
+origin  https://github.com/<your_username>/qmk_firmware.git (push)
+upstream        https://github.com/qmk/qmk_firmware.git (fetch)
+upstream        https://github.com/qmk/qmk_firmware.git (push)
+```
+
+Now that this is done, you can check for updates to the repository by running `git fetch upstream`. This retrieves the list of commits to QMK's `master` branch- nicknamed `upstream` - which then can be compared to your own fork, nicknamed `origin`. To update your fork's master, run:
+
+```
+git checkout master
+git pull upstream master
```

Comment:
You should have the `git fetch upstream` here.

I know you mentioned it above, but it should be here explicitly.  

Without it, tags won't get updated properly, for instance.

Reply:
Also, this is rebasing.... may want to add `--rebase` to the pull?  Not sure if it's 100% needed.

But what we do need is a section on conflict resolution.   Because it's not if, but ***when*** a merge conflict will happen. 

`git status` is great for finding what file it is.  But a GUI is helpful for finding the merge conflicts. 

And `git add . && git rebase --continue` to add the changes and continue rebasing.

Reply:
I've always been under the impression that `git pull` runs a `git fetch` automatically. I nearly never use a `git fetch` on its own and I've never had any unintended behavior.

I was actually trying to avoid getting into `git rebase` in this section, as I feel it's a bit advanced for someone brand new to Git. For a later section, or maybe another page if this turns into a subsection.

Reply:
it does, mostly, but it does not pull down the tags when it does this.  To get the tags (eg 0.6.100), you need to run fetch first.

Reply:
I tested it before I replied.

`git hist` is an alias for a `--pretty`-formatted `git log`. The `git pull` operation pulls down two PRs (3868 and 3879), and the tags are present in the second pass of `git hist`.

https://gist.github.com/noroadsleft/162887cf995798d4642187550b66c3b2

Reply:
So... unless I'm reading that wrong, yes, we DO want the `git fetch` first.

If I'm wrong.... well, I'm not.  I have it in my script because I had issues with that. And I'm not the only one that has....  So a `git fetch upstream` before the pull/rebase is going to be "best practice"

Reply:
if I had my aliases in git.... I'd show you that I was using `git pull --rebase upstream master` for a while.  But it NEVER got the tags.   I ***had*** to add `git fetch` before the pull command to get it to ACTUALLY get the tags. 

So, there is something in there that isn't quite right. Or something with how git on MSYS works. 

But I'm still saying that myself and others have HAD to use `git fetch upstream` to properly update the tags.

Reply:
Maybe it's the `--rebase` argument, which I don't think I've ever used.

Come to think of it, we might be talking about two different tasks here. The goal for this section is only about how to update the `master` branch. No need for `--rebase` here because we're not rebasing anything.

Though this doesn't do anything for dealing with merge conflicts, which I want to cover in a separate section or document (as part of a series).

Reply:
Just tested this.  Does NOT pull the tags:

```
48a992f1c037658bbacccefd2709ffdcda8bb345 (HEAD -> master, upstream/master) Zeal60/Zeal65/M60-A implementation (#3879)
170de1273cde78d93b3f955aca133a4d690742b2 Add an easy way to create new keymaps for your favorite keyboard (#3868)
6d6d91c834ef3415425e21d895d4ec91c67fd4b8 (tag: 0.6.117, origin/master, origin/HEAD) rgblight.[ch] more configurable (#3582)
```

Reply:
Full output: https://gist.github.com/drashna/e414cc7a76bb7d98e87c32a27ff0baea

Notice how it only even showed the tags when I did `git fetch upstream`?  it ignored them otherwise.

Reply:
I wonder if this is Git Bash via Windows, or the fact that I'm using `--graph`. Those are the only reasons I can think that mine would work differently.

I'll be changing this.